configuration/config2/ConfigItem.hpp

```diff
@@ -0,0 +1,148 @@
+
+        virtual bool addUniqueName(const std::string keyName);
+        virtual void addUniqueAttributeValueSetDefinition(const std::string &setName, const std::string &elementPath, const std::string &attributeName, bool duplicateOk = false);
+        virtual void addReferenceToUniqueAttributeValueSet(const std::string &setName, const std::string &elementPath, const std::string &attributeName);
+        virtual void processUniqueAttributeValueSets();
+        virtual void processUniqueAttributeValueSetReferences();
+
+        virtual void resetEnvironment(); 
+
+        virtual void postProcessConfig();
+
+        ConfigItem() { };
+
+
+    protected:
+
+        std::string m_name;
+        std::string m_displayName;
+        std::string m_className;
+        std::string m_category;  // used for further subdividing to the user
+        std::string m_componentName;   
+        std::multimap<std::string, std::shared_ptr<ConfigItem>> m_children; 
+        std::shared_ptr<CfgValue> m_pItemCfgValue;   // value for this item (think of it as the VALUE for an element <xx attr= att1=>VALUE</xx>)
+        std::map<std::string, std::shared_ptr<CfgValue>> m_attributes;   // attributes for this item (think in xml terms <m_name attr1="val" attr2="val" .../> where attrN is in this vector
+        std::set<std::string> m_keys;   // generic set of key values for use by any component to prevent duplicate operations
+        std::weak_ptr<ConfigItem> m_pParent;
+
+        std::map<std::string, std::shared_ptr<CfgType>> m_types;
+        std::map<std::string, std::shared_ptr<ConfigItem>> m_configTypes;                // reusable types
+
+        int m_minInstances;
+        int m_maxInstances;
+        int m_version;
+
+        // This struct handles both unique attribute sets and references to same
+        struct SetInfo {
+            SetInfo(const std::string &setName, const std::string &elementPath, const std::string &attributeName) :
+                m_setName(setName), m_elementPath(elementPath), m_attributeName(attributeName), m_duplicateOk(false) { }
+            SetInfo(const std::string &setName, const std::string &elementPath, const std::string &attributeName, bool duplicateOk) :
+                m_setName(setName), m_elementPath(elementPath), m_attributeName(attributeName), m_duplicateOk(duplicateOk) { }
+            std::string m_setName; 
+            std::string m_elementPath; 
+            std::string m_attributeName;
+            bool m_duplicateOk;
+        };
+
+        // Attribute unique sets and references to unique sets are stored during parsing and post processed
+        std::map<std::string, SetInfo> m_uniqueAttributeValueSetReferences;
+        std::map<std::string, SetInfo> m_uniqueAttributeValueSetDefs;
+
+        // These are the attribute value sets whose members must be unique
+        static std::map<std::string, std::vector<std::shared_ptr<CfgValue>>> m_uniqueAttributeValueSets;    
```

Comment:
are std:maps thread safe? if not does this static map need to be thread access controlled?